alibi/explainers/cfrl_tabular.py

```diff
@@ -0,0 +1,310 @@
+        if "consistency_loss" not in kwargs:
+            self.params["consistency_loss"] = self.backend.consistency_loss
+
+        # Set training conditional function generator if not user-specified.
+        if "conditional_func" not in kwargs:
+            self.params["conditional_func"] = partial(self.backend.generate_condition,
+                                                      feature_names=self.params["feature_names"],
+                                                      category_map=self.params["category_map"],
+                                                      ranges=self.params["ranges"],
+                                                      immutable_features=self.params["immutable_features"])
+
+        # Set testing conditional function generator if not user-specified.
+        if "conditional_vector" not in kwargs:
+            self.params["conditional_vector"] = partial(conditional_vector,
+                                                        preprocessor=self.params["ae_preprocessor"],
+                                                        feature_names=self.params["feature_names"],
+                                                        category_map=self.params["category_map"],
+                                                        ranges=self.params["ranges"],
+                                                        immutable_features=self.params["immutable_features"])
+
+    def _select_backend(self, backend, **kwargs):
+        """
+        Selects the backend according to the `backend` flag.
+
+        Parameters
+        ----------
+        backend
+            Deep learning backend. `tensorflow`|`pytorch`. Default `tensorflow`.
+        """
+        return tensorflow_tabular_backend if backend == "tensorflow" else pytorch_tabular_backend
+
+    def fit(self, x: np.ndarray) -> 'Explainer':
+        # Compute vector of statistics to clamp numerical values between the minimum and maximum
+        # value from the training set.
+        self.params["stats"] = statistics(x=x,
+                                          preprocessor=self.params["ae_preprocessor"],
+                                          category_map=self.params["category_map"])
+
+        # Set postprocessing functions. Needs `stats`.
+        self.params["postprocessing_funcs"] = [
+            SamplePostprocessing(stats=self.params["stats"], category_map=self.params["category_map"]),
+            ConcatPostprocessing(),
+        ]
+
+        return super().fit(x)
+
+    def explain(self,
+                X: np.ndarray,
+                y_t: np.ndarray = None,  # TODO remove default value (mypy error)
+                c: Optional[List[Dict[str, List[Union[str, float]]]]] = None,
+                diversity: bool = False,
+                num_samples: int = 1,
+                batch_size: int = 100,
+                patience: int = 1000,
+                tolerance: float = 1e-3) -> Explanation:
```

Comment:
The CF-RL has normally a deterministic behavior. For the tabular case we are able to generate a diverse counterfactual set by generating conditions that are a subset of the original condition. In this way, we keep the feasibility of the counterfactual. To actually generate a subset of the original condition, you need to know some insights regarding the condition vector. It can probably be extended, but I think it might require some user-defined subset condition generator.  At the same time, it can be a bit difficult since the explain method in the base, expects a numerical conditional vector which is not really human-readable.

Reply:
OK, I wasn't really suggesting extending the base class to handle `diversity` more that e.g. `batch_size` seemed like a general parameter that's not present in `cfrl_base`  ?

Reply:
I've already included the `batch_size` in the following PR.